# ============================================================================
# ALERTS REPOSITORY
# ============================================================================
# This file was auto-generated on: 2025-09-30 13:48:44 WAT
# It contains asynchronous functions for managing access to the database
# in a MongoDB database using FastAPI.
#
# DO NOT EDIT THIS FILE MANUALLY - RE-RUN THE GENERATOR INSTEAD. OR IF YOU WANT TO EDIT JUST ADD LEAVE OTHER FUNCTIONS THE WAY YOU MET THEM
# ============================================================================

from pymongo import ReturnDocument
from core.database import db
from fastapi import HTTPException,status
from typing import List,Optional
from schemas.alerts import AlertsUpdate, AlertsCreate, AlertsOut

async def create_alerts(alerts_data: AlertsCreate) -> AlertsOut:
    alerts_dict = alerts_data.model_dump()
    result =await db.alertss.insert_one(alerts_dict)
    result = await db.alertss.find_one(filter={"_id":result.inserted_id})
    returnable_result = AlertsOut(**result)
    return returnable_result

async def get_alerts(filter_dict: dict) -> Optional[AlertsOut]:
    try:
        result = await db.alertss.find_one(filter_dict)

        if result is None:
            return None

        return AlertsOut(**result)

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"An error occurred while fetching alerts: {str(e)}"
        )
    
async def get_alertss(filter_dict: dict = {},start=0,stop=100) -> List[AlertsOut]:
    try:
        if filter_dict is None:
            filter_dict = {}

        cursor = (db.alertss.find(filter_dict)
        .skip(start)
        .limit(stop - start)
        )
        alerts_list = []

        async for doc in cursor:
            alerts_list.append(AlertsOut(**doc))

        return alerts_list

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"An error occurred while fetching alertss: {str(e)}"
        )
async def update_alerts(filter_dict: dict, alerts_data: AlertsUpdate) -> AlertsOut:
    result = await db.alertss.find_one_and_update(
        filter_dict,
        {"$set": alerts_data.model_dump(exclude_none=True)},
        return_document=ReturnDocument.AFTER
    )
    returnable_result = AlertsOut(**result)
    return returnable_result

async def delete_alerts(filter_dict: dict):
    return await db.alertss.delete_one(filter_dict)